### PR TITLE
Move `get_valid_dir_option` to module method

### PR DESCRIPTION
**Pull request by portante**
We move the `PbenchServerConfig.get_valid_dir_option()` method to a module method, renaming it to `get_resolved_dir()`.  We also move `_get_valid_path` to a module method.

### CODE REVIEW

**Review by dbutenhof — APPROVED** [review]

**Review by riya-17 — DISMISSED** [review]

**Review by npalaska — APPROVED** [review]
Looks okay, but I have a question; Most of the places where we are using the `_get_valid_path` is to get a fully resolved path of a directory either specifically created when building the server image like `results` `incoming` or using the server config file like `backup`, etc. Arent all these parts of server configuration, so why the need to move it outside the `PbenchServerConfig` class?

File: lib/pbench/server/__init__.py [hunk 1]
```diff
@@ -65,6 +65,67 @@ def tstos(ts: float = None) -> str:
...
+
+    Returns:
+        A Path directory object, None if the directory value does not
+            resolve to a real directory.
+    """
+    dir_path = _get_valid_path(env_name, dir_val, logger)
+    if not dir_path:
+        return None
+
+    return dir_path
```
Comment (webbnh) [inline lib/pbench/server/__init__.py:126]: Does this code accomplish anything?

The only times that `not dir_path` is `True` is when `dir_path` is `None`, so line 126 would produce the same result if we removed lines 123-125.  And, so `get_resolved_dir()` appears to be a straight pass-through for `_get_valid_path()` (i.e., it it doesn't rename or reorder the parameters).

Can't we just rename `_get_valid_path()` to `get_resolved_dir()` and remove this function?

**Review by webbnh — APPROVED** [review]
Looks OK, but I think it would be better without what appears to me to be a pointless wrapper function.

**Comment by portante** [comment]
> Looks okay, but I have a question; Most of the places where we are using the `_get_valid_path` is to get a fully resolved path of a directory either specifically created when building the server image like `results` `incoming` or using the server config file like `backup`, etc. Arent all these parts of server configuration, so why the need to move it outside the `PbenchServerConfig` class?

The `PbenchServerConfig` class is about fetching configuration information.  Validating directory strings is context sensitive where that context is not known by `PbenchServerConfig` in all cases.

File: lib/pbench/server/__init__.py (hunk 1 above)
Comment (portante) [reply lib/pbench/server/__init__.py:126]: > Can't we just rename `_get_valid_path()` to `get_resolved_dir()` and remove this function?

It looks like we can.

Comment (webbnh) [reply lib/pbench/server/__init__.py:126]: Yes, but _should_ we?  _**Will**_ we??  (Inquiring minds want to know! 😆)

**Review by portante — COMMENTED** [review]
Applied code review comments.

File: lib/pbench/server/__init__.py (hunk 1 above)
Comment (portante) [reply lib/pbench/server/__init__.py]: No ad revenue if I give away the ending.

**Review by webbnh — APPROVED** [review]
👍

**Review by npalaska — APPROVED** [review]

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/server/__init__.py (hunk 1 above)
Comment (npalaska) [reply lib/pbench/server/__init__.py]: :rofl:

**Comment by portante** [comment]
@npalaska, functional tests failed in a weird way:

```
pbench-in-a-can_3227-pbenchserver
b6d01cc3fd6f0c6154fbf794e1b9fd1c9afd662803970655c4765b1f48a5730d
b6d01cc3fd6f0c6154fbf794e1b9fd1c9afd662803970655c4765b1f48a5730d
time="2023-02-03T20:21:44Z" level=error msg="Removing container: failed to clean up and remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4: cannot remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4 as it is running - running or paused containers cannot be removed without force: container state improper"
Error: failed to clean up and remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4: cannot remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4 as it is running - running or paused containers cannot be removed without force: container state improper
```

**Comment by dbutenhof** [comment]
> @npalaska, functional tests failed in a weird way:
> 
> ```
> pbench-in-a-can_3227-pbenchserver
> b6d01cc3fd6f0c6154fbf794e1b9fd1c9afd662803970655c4765b1f48a5730d
> b6d01cc3fd6f0c6154fbf794e1b9fd1c9afd662803970655c4765b1f48a5730d
> time="2023-02-03T20:21:44Z" level=error msg="Removing container: failed to clean up and remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4: cannot remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4 as it is running - running or paused containers cannot be removed without force: container state improper"
> Error: failed to clean up and remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4: cannot remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4 as it is running - running or paused containers cannot be removed without force: container state improper
> ```

Not just that, but

> Feb 03 20:21:35 pbenchinacan pbench-server[99]: [2023-02-03 20:21:35 +0000] [99] [ERROR] Can't connect to ('0.0.0.0', 8001)
Feb 03 20:21:35 pbenchinacan gunicorn.pbench-server.error[99]: [99] Can't connect to ('0.0.0.0', 8001)
Feb 03 20:21:35 pbenchinacan systemd[1]: pbench-server.service: Main process exited, code=exited, status=1/FAILURE
Feb 03 20:21:35 pbenchinacan systemd[1]: pbench-server.service: Failed with result 'exit-code'.

**Comment by npalaska** [comment]
> pbench-in-a-can_3227-pbenchserver
> b6d01cc3fd6f0c6154fbf794e1b9fd1c9afd662803970655c4765b1f48a5730d
> b6d01cc3fd6f0c6154fbf794e1b9fd1c9afd662803970655c4765b1f48a5730d
> time="2023-02-03T20:21:44Z" level=error msg="Removing container: failed to clean up and remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4: cannot remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4 as it is running - running or paused containers cannot be removed without force: container state improper"
> Error: failed to clean up and remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4: cannot remove container 6c2bfd1668a2e338a793715411d42febd2bd75c11ff3363a8bbb6bee8afb39d4 as it is running - running or paused containers cannot be removed without force: container state improper

This is really weird it has tried to stop the running pbench-server container and it failed. But this is the first time I am seeing this type of failure. Wonder if there was a running container left around from earlier? I think probably best to force stop the pbench-server container at the end of `run-server-func-test`.

**Comment by npalaska** [comment]
`test_upload` also failed because of an internal server error. This might be just a very isolated occurrence, we can try to rebuild it I guess and see if it goes away.

**Comment by dbutenhof** [comment]
> `test_upload` also failed because of an internal server error. This might be just a very isolated occurrence, we can try to rebuild it I guess and see if it goes away.

gunicorn never started up because it couldn't bind `0.0.0.0:8001` for some reason... in fact it looks like the nginx had the same trouble with `8080`...

Yeah, this is a completely new failure mode. Congratulations, someone! 😆 🥇 (OK, that's a medal; but it looks like a pineapple with blue foliage...)

**Comment by dbutenhof** [comment]
And looking through the log, I realize that the new Keycloak configuration generates a *lot* of output, including auth token and secret key. Can we quiet this down a bit, maybe?

```
Waiting for the Keycloak server
Waiting for the Keycloak server
Waiting for the Keycloak server
Waiting for the Keycloak server
Waiting for the Keycloak server
Waiting for the Keycloak server
Waiting for the Keycloak server
Keycloak health url is up
Keycloak Server is up

KEYCLOAK_HOST_PORT: http://localhost:8090/

Getting admin access token
--------------------------
ADMIN_TOKEN=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJrYVF4MFIxS04zaXgyVGhNaEJhOHBna3RsdGlwSTFEQ2E4Q1BOcUwwZl9jIn0.eyJleHAiOjE2NzU0NTU3MDksImlhdCI6MTY3NTQ1NTY0OSwianRpIjoiODU1NWFlMGYtMmUxZS00YjVhLThmNDgtZWJmMDQzODU1ZWE4IiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo4MDkwL3JlYWxtcy9tYXN0ZXIiLCJzdWIiOiJmYjdkMzkxYi1mZDg1LTRhZWMtYjExMC1lZWQ2ODRkMDM3ODciLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJhZG1pbi1jbGkiLCJzZXNzaW9uX3N0YXRlIjoiMTFhNDY3ZDctMmU1OC00OTU4LTlhMDAtZTU5ZTY1Y2VmZGRlIiwiYWNyIjoiMSIsInNjb3BlIjoicHJvZmlsZSBlbWFpbCIsInNpZCI6IjExYTQ2N2Q3LTJlNTgtNDk1OC05YTAwLWU1OWU2NWNlZmRkZSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwicHJlZmVycmVkX3VzZXJuYW1lIjoiYWRtaW4ifQ.efUg4Ee-FF7HhO056CK0q_0hVFXB2t2gwVf7XJvrrVSyYc9TlvQPOtzTPvLIp6-o89fmEgufzu7-BdVlei7xc67kUjDFCHCVWj9UiZ9vY4DNYNmwfgsV2dCdLVAPmPZH7Oy01XXDubXjHREo5Sp-501o2cI7eIT2DulB_7HjnE3SrrF2oTE4kGsYBYvNtmQ3KZ5STildjIeTQa7-Etd-Dq8TgE0vPWxV9X4SK5J1NtjKf7dMOVGfzWaiPzLwedMFCGGK9fAEdvsOg6qpm5DbGjSfo1Fb2JkkwD9H9G6m4Gh54lfmkKpEw3txY0BuKFrG_00WHKD4-bc0gYsLI3wO0A

Creating pbench realm
--------------
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    36    0     0  100    36      0    180 --:--:-- --:--:-- --:--:--   179
100    36    0     0  100    36      0     44 --:--:-- --:--:-- --:--:--    43
Creating pbench-server client
---------------
client_id=d00f50d3-8712-4405-aee1-9444f6d018d3

Getting client secret
---------------------
PBENCH_CLIENT_SECRET=KKF2KvgHTu4JXYbtzXLQE3zZSgZcCpbI

Creating an 'ADMIN' role under pbench-server client of the pbench realm
--------------------
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    17    0     0  100    17      0    809 --:--:-- --:--:-- --:--:--   809
ROLE_ID=28e7268a-010a-4b0f-82dc-6ec38ac5b76e

Creating an 'admin' user inside pbench realm
-------------
USER_ID=3b7a3dd5-7b8c-41c4-b497-a4e65b83068e

Assigning 'ADMIN' client role to the user 'admin' created above
---------------------------
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    62    0     0  100    62      0   2952 --:--:-- --:--:-- --:--:--  2952
The Keycloak configuration is complete.
```

**Comment by npalaska** [comment]
@dbutenhof I am working on that, I already raised a PR I am just updating the PR description and all, it should land in couple mins.

**Pull request merged**